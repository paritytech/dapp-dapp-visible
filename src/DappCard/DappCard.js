// Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import React, { Component } from 'react';
import { observer } from 'mobx-react';
import PropTypes from 'prop-types';
import Card from 'semantic-ui-react/dist/commonjs/views/Card';
import Button from 'semantic-ui-react/dist/commonjs/elements/Button';
import Label from 'semantic-ui-react/dist/commonjs/elements/Label';
import DappIcon from '@parity/ui/lib/DappIcon';
import { FormattedMessage } from 'react-intl';

import DappVouchFor from './DappVouchFor';
import styles from './DappCard.css';

const DAPP_DAPP_VISIBLE_ID =
  '0xa48bd8fd56c90c899135281967a6cf90865c221b46f27f9fbe2a236d74a64ea2'; // This dapp's id

class DappCard extends Component {
  static propTypes = {
    dapp: PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      image: PropTypes.string
    }).isRequired,
    onAdd: PropTypes.func.isRequired,
    onOpen: PropTypes.func.isRequired,
    onRemove: PropTypes.func.isRequired,
    visible: PropTypes.bool
  };

  handleAdd = () => this.props.onAdd(this.props.dapp.id);

  handleRemove = () => this.props.onRemove(this.props.dapp.id);

  handleOpen = () => this.props.onOpen(this.props.dapp.id);

  render() {
    const { dapp, visible } = this.props;

    return (
      <Card>
        <Card.Content>
          <DappIcon
            app={dapp}
            className={`ui centered image ${styles.picture}`}
          />
          <Card.Header>{dapp.name}</Card.Header>
          <Card.Meta>{dapp.description}</Card.Meta>
          <Card.Description>
            {dapp.type === 'local' && (
              <Label>
                <FormattedMessage
                  id="dapp.visible.localLabel"
                  defaultMessage="Local Dapp"
                />
              </Label>
            )}
            {dapp.type === 'builtin' && (
              <Label>
                <FormattedMessage
                  id="dapp.visible.builtinLabel"
                  defaultMessage="By Parity"
                />
              </Label>
            )}
            <DappVouchFor dapp={dapp} />
          </Card.Description>
        </Card.Content>
        <Card.Content extra>
          <Button
            basic
            secondary
            content={
              <FormattedMessage
                id="dapp.visible.openDapp"
                defaultMessage="Open"
              />
            }
            icon="external"
            onClick={this.handleOpen}
          />
          {visible ? (
            <Button
              content={
                <FormattedMessage
                  id="dapp.visible.addedDapp"
                  defaultMessage="Added to home"
                />
              }
              disabled={dapp.id === DAPP_DAPP_VISIBLE_ID} // Do not allow removing this dapp from homepage
              icon="toggle on"
              onClick={this.handleRemove}
              primary
            />
          ) : (
            <Button
              basic
              content={
                <FormattedMessage
                  id="dapp.visible.addDapp"
                  defaultMessage="Add to home"
                />
              }
              icon="toggle off"
              onClick={this.handleAdd}
              primary
            />
          )}
        </Card.Content>
      </Card>
    );
  }
}

export default observer(DappCard);
