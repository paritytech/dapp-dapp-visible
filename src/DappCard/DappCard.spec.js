// Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import React from 'react';
import { shallowToJson } from 'enzyme-to-json';
import Button from 'semantic-ui-react/dist/commonjs/elements/Button';

import { shallowWithIntl } from '../setupTests';
import DappCard from './DappCard';

// Mock props
const props = {
  dapp: {
    id: '123',
    name: 'Test'
  },
  visible: false,
  onAdd: () => {},
  onOpen: () => {},
  onRemove: () => {}
};

test('should render correctly when visible', () => {
  const component = shallowWithIntl(<DappCard {...props} />);

  expect(shallowToJson(component)).toMatchSnapshot();
});

test('should render correctly when hidden', () => {
  const component = shallowWithIntl(<DappCard {...props} visible />);

  expect(shallowToJson(component)).toMatchSnapshot();
});

test('should render correctly a local dapp', () => {
  const component = shallowWithIntl(
    <DappCard {...props} dapp={{ ...props.dapp, type: 'local' }} />
  );

  expect(shallowToJson(component)).toMatchSnapshot();
});

test('should render correctly a builtin dapp', () => {
  const component = shallowWithIntl(
    <DappCard {...props} dapp={{ ...props.dapp, type: 'builtin' }} />
  );

  expect(shallowToJson(component)).toMatchSnapshot();
});

test('should handle onOpen on Open button click', () => {
  const onOpen = jest.fn();

  const component = shallowWithIntl(<DappCard {...props} onOpen={onOpen} />);

  component
    .find(Button)
    .first()
    .simulate('click');
  expect(onOpen).toHaveBeenCalledWith('123');
});

test('should handle onAdd on Add To Home button click', () => {
  const onAdd = jest.fn();

  const component = shallowWithIntl(<DappCard {...props} onAdd={onAdd} />);

  component
    .find(Button)
    .at(1)
    .simulate('click');
  expect(onAdd).toHaveBeenCalledWith('123');
});

test('should handle onRemove on Remove button click', () => {
  const onRemove = jest.fn();

  const component = shallowWithIntl(
    <DappCard {...props} onRemove={onRemove} visible />
  );

  component
    .find(Button)
    .last()
    .simulate('click');
  expect(onRemove).toHaveBeenCalledWith('123');
});
