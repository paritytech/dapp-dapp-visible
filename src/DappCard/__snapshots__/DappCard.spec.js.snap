// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly a builtin dapp 1`] = `
<Card>
  <CardContent>
    <DappIcon
      app={
        Object {
          "id": "123",
          "name": "Test",
          "type": "builtin",
        }
      }
      className="ui centered image undefined"
      raised={true}
      size="normal"
    />
    <CardHeader>
      Test
    </CardHeader>
    <CardMeta />
    <CardDescription>
      <Label>
        <FormattedMessage
          defaultMessage="By Parity"
          id="dapp.visible.builtinLabel"
          values={Object {}}
        />
      </Label>
      <DappVouchFor
        dapp={
          Object {
            "id": "123",
            "name": "Test",
            "type": "builtin",
          }
        }
      />
    </CardDescription>
  </CardContent>
  <CardContent
    extra={true}
  >
    <Button
      as="button"
      basic={true}
      content={
        <FormattedMessage
          defaultMessage="Open"
          id="dapp.visible.openDapp"
          values={Object {}}
        />
      }
      icon="external"
      onClick={[Function]}
      secondary={true}
    />
    <Button
      as="button"
      basic={true}
      content={
        <FormattedMessage
          defaultMessage="Add to home"
          id="dapp.visible.addDapp"
          values={Object {}}
        />
      }
      icon="toggle off"
      onClick={[Function]}
      primary={true}
    />
  </CardContent>
</Card>
`;

exports[`should render correctly a local dapp 1`] = `
<Card>
  <CardContent>
    <DappIcon
      app={
        Object {
          "id": "123",
          "name": "Test",
          "type": "local",
        }
      }
      className="ui centered image undefined"
      raised={true}
      size="normal"
    />
    <CardHeader>
      Test
    </CardHeader>
    <CardMeta />
    <CardDescription>
      <Label>
        <FormattedMessage
          defaultMessage="Local Dapp"
          id="dapp.visible.localLabel"
          values={Object {}}
        />
      </Label>
      <DappVouchFor
        dapp={
          Object {
            "id": "123",
            "name": "Test",
            "type": "local",
          }
        }
      />
    </CardDescription>
  </CardContent>
  <CardContent
    extra={true}
  >
    <Button
      as="button"
      basic={true}
      content={
        <FormattedMessage
          defaultMessage="Open"
          id="dapp.visible.openDapp"
          values={Object {}}
        />
      }
      icon="external"
      onClick={[Function]}
      secondary={true}
    />
    <Button
      as="button"
      basic={true}
      content={
        <FormattedMessage
          defaultMessage="Add to home"
          id="dapp.visible.addDapp"
          values={Object {}}
        />
      }
      icon="toggle off"
      onClick={[Function]}
      primary={true}
    />
  </CardContent>
</Card>
`;

exports[`should render correctly when hidden 1`] = `
<Card>
  <CardContent>
    <DappIcon
      app={
        Object {
          "id": "123",
          "name": "Test",
        }
      }
      className="ui centered image undefined"
      raised={true}
      size="normal"
    />
    <CardHeader>
      Test
    </CardHeader>
    <CardMeta />
    <CardDescription>
      <DappVouchFor
        dapp={
          Object {
            "id": "123",
            "name": "Test",
          }
        }
      />
    </CardDescription>
  </CardContent>
  <CardContent
    extra={true}
  >
    <Button
      as="button"
      basic={true}
      content={
        <FormattedMessage
          defaultMessage="Open"
          id="dapp.visible.openDapp"
          values={Object {}}
        />
      }
      icon="external"
      onClick={[Function]}
      secondary={true}
    />
    <Button
      as="button"
      content={
        <FormattedMessage
          defaultMessage="Added to home"
          id="dapp.visible.addedDapp"
          values={Object {}}
        />
      }
      disabled={false}
      icon="toggle on"
      onClick={[Function]}
      primary={true}
    />
  </CardContent>
</Card>
`;

exports[`should render correctly when visible 1`] = `
<Card>
  <CardContent>
    <DappIcon
      app={
        Object {
          "id": "123",
          "name": "Test",
        }
      }
      className="ui centered image undefined"
      raised={true}
      size="normal"
    />
    <CardHeader>
      Test
    </CardHeader>
    <CardMeta />
    <CardDescription>
      <DappVouchFor
        dapp={
          Object {
            "id": "123",
            "name": "Test",
          }
        }
      />
    </CardDescription>
  </CardContent>
  <CardContent
    extra={true}
  >
    <Button
      as="button"
      basic={true}
      content={
        <FormattedMessage
          defaultMessage="Open"
          id="dapp.visible.openDapp"
          values={Object {}}
        />
      }
      icon="external"
      onClick={[Function]}
      secondary={true}
    />
    <Button
      as="button"
      basic={true}
      content={
        <FormattedMessage
          defaultMessage="Add to home"
          id="dapp.visible.addDapp"
          values={Object {}}
        />
      }
      icon="toggle off"
      onClick={[Function]}
      primary={true}
    />
  </CardContent>
</Card>
`;
